rskj-core/src/main/java/co/rsk/peg/RepositoryBtcBlockStoreWithCache.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2017 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.peg;
+
+import co.rsk.bitcoinj.core.BtcBlock;
+import co.rsk.bitcoinj.core.NetworkParameters;
+import co.rsk.bitcoinj.core.Sha256Hash;
+import co.rsk.bitcoinj.core.StoredBlock;
+import co.rsk.bitcoinj.store.BlockStoreException;
+import co.rsk.core.RskAddress;
+import co.rsk.util.MaxSizeHashMap;
+import org.ethereum.core.Repository;
+import org.ethereum.vm.DataWord;
+import org.ethereum.vm.PrecompiledContracts;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+
+/**
+ * Implementation of a bitcoinj blockstore that persists to RSK's Repository
+ * @author Oscar Guindzberg
+ */
+public class RepositoryBtcBlockStoreWithCache implements BtcBlockStoreWithCache {
+    public static final String BLOCK_STORE_CHAIN_HEAD_KEY = "blockStoreChainHead";
+    private final Repository repository;
+    private final RskAddress contractAddress;
+    private final NetworkParameters btcNetworkParams;
+    public static final int MAX_DEPTH_STORED_BLOCKS = 5_000;
+    public static final int MAX_SIZE_MAP_STORED_BLOCKS = 50_000;
```

Comment:
This value is higher than what is necessary in practice. Check if the initial capacity (50000) and load factor (0.75)  of `LinkedHashMap` do what you expect for this cache. Ideally only `MAX_DEPTH_STORED_BLOCKS` plus a few parallel blocks should be kept in memory and the minimum size of the data structure should not grow.

Reply:
Lowered the number to 10_000